keel-core/src/main/kotlin/com/netflix/spinnaker/keel/core/api/ClusterDeployStrategy.kt

```diff
@@ -23,42 +23,90 @@ import java.time.Duration.ZERO
+  abstract fun withDefaultsOmitted(): ClusterDeployStrategy
 }
 
 @JsonTypeName("red-black")
 data class RedBlack(
-  val rollbackOnFailure: Boolean = true,
-  val resizePreviousToZero: Boolean = false,
-  val maxServerGroups: Int = 2,
-  val delayBeforeDisable: Duration = ZERO,
-  val delayBeforeScaleDown: Duration = ZERO,
+  val rollbackOnFailure: Boolean? = true,
+  val resizePreviousToZero: Boolean? = false,
+  val maxServerGroups: Int? = 2,
+  val delayBeforeDisable: Duration? = ZERO,
+  val delayBeforeScaleDown: Duration? = ZERO,
   // The order of this list is important for pauseTime based staggers
   @JsonInclude(NON_EMPTY)
   override val stagger: List<StaggeredRegion> = emptyList()
 ) : ClusterDeployStrategy() {
+
+  companion object {
+    fun fromOrcaStageContext(context: Map<String, Any?>) =
+      RedBlack(
+        rollbackOnFailure = context["rollback"]
+          ?.let { it as Map<String, Any> }
+          ?.get("onFailure") as Boolean,
+        resizePreviousToZero = context["scaleDown"] as Boolean,
+        maxServerGroups = context["maxRemainingAsgs"] as Int,
+        delayBeforeDisable = Duration.ofSeconds((context["delayBeforeDisableSec"] as Int).toLong()),
+        delayBeforeScaleDown = Duration.ofSeconds((context["delayBeforeScaleDownSec"] as Int).toLong())
+      )
+
+    val DEFAULTS = RedBlack()
+  }
+
   override fun toOrcaJobProperties() = mapOf(
     "strategy" to "redblack",
     "maxRemainingAsgs" to maxServerGroups,
-    "delayBeforeDisableSec" to delayBeforeDisable.seconds,
-    "delayBeforeScaleDownSec" to delayBeforeScaleDown.seconds,
+    "delayBeforeDisableSec" to delayBeforeDisable?.seconds,
+    "delayBeforeScaleDownSec" to delayBeforeScaleDown?.seconds,
     "scaleDown" to resizePreviousToZero,
     "rollback" to mapOf("onFailure" to rollbackOnFailure)
   )
 
   override val isStaggered: Boolean
-    get() = stagger.isNotEmpty()
+    get() = stagger?.isNotEmpty() ?: false
+
+  override fun withDefaultsOmitted() =
+    RedBlack(
+      maxServerGroups = if (maxServerGroups == DEFAULTS.maxServerGroups) {
+        null
+      } else {
+        maxServerGroups
+      },
+      delayBeforeDisable = if (delayBeforeDisable == DEFAULTS.delayBeforeDisable) {
+        null
+      } else {
+        delayBeforeDisable
+      },
+      delayBeforeScaleDown = if (delayBeforeScaleDown == DEFAULTS.delayBeforeScaleDown) {
+        null
+      } else {
+        delayBeforeScaleDown
+      },
+      resizePreviousToZero = if (resizePreviousToZero == DEFAULTS.resizePreviousToZero) {
+        null
+      } else {
+        resizePreviousToZero
+      },
+      rollbackOnFailure = if (rollbackOnFailure == DEFAULTS.rollbackOnFailure) {
+        null
+      } else {
+        rollbackOnFailure
+      }
+    )
 }
 
 @JsonTypeName("highlander")
 object Highlander : ClusterDeployStrategy() {
   override fun toOrcaJobProperties() = mapOf(
     "strategy" to "highlander"
   )
+
+  override fun withDefaultsOmitted() = this
```

Comment:
So the `=` part is just a shorthand-form for a return type + function body when the return type can be inferred from the expression following (`this`, in this case). The `this` part is because, in this case, `Highlander` is not actually a class, but an `object` (singleton instance), and it has no properties. If it was a class, `this` would be replaced with `Highlander()` to instantiate an object. Make sense?

Reply:
Yeah, makes sense. That was a little mind-bendy.